[GTK][WPE] Integrate the ports with MALLOC_HEAP_BREAKDOWN

Scony opened the pull request:
#### 6599b99887279f51eefab92d6cc9be0db2e700f2
<pre>
[GTK][WPE] Integrate the ports with MALLOC_HEAP_BREAKDOWN
<a href="https://bugs.webkit.org/show_bug.cgi?id=291063">https://bugs.webkit.org/show_bug.cgi?id=291063</a>

Reviewed by Justin Michaud and Adrian Perez de Castro.

This change integrates GTK and WPE ports with MALLOC_HEAP_BREAKDOWN
feature.

Despite the MALLOC_HEAP_BREAKDOWN was meant to be apple-port only,
it&apos;s fairly straightforward for Glib-based ports to export the
malloc.h header along with simple library that implements a mechanism
for tracking the allocations. With that, it&apos;s possible to gather
statistics on the current allocations at any point in time.

This change lays a groundwork for more sophisticated integration by
providing a simplified malloc.h header along with a simple library
that allows logging statistics to the standard output periodically.

Co-authored-by: Justin Michaud &lt;jmichaud@igalia.com&gt;

* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/DebugHeap.cpp:
* Source/WTF/wtf/DebugHeap.h:
* Source/WTF/wtf/MallocHeapBreakdown.cmake: Added.
* Source/WTF/wtf/malloc_heap_breakdown/main.cpp: Added.
(MallocZoneHeapManager::getInstance):
(MallocZoneHeapManager::defaultZone):
(MallocZoneHeapManager::createZone):
(MallocZoneHeapManager::renameZone):
(MallocZoneHeapManager::zoneMalloc):
(MallocZoneHeapManager::zoneCalloc):
(MallocZoneHeapManager::zoneRealloc):
(MallocZoneHeapManager::zoneMemalign):
(MallocZoneHeapManager::zoneFree):
(MallocZoneHeapManager::MallocZoneHeapManager):
(MallocZoneHeapManager::~MallocZoneHeapManager):
(MallocZoneHeapManager::monitoringThreadMain):
(malloc_default_zone):
(malloc_create_zone):
(malloc_zone_malloc):
(malloc_zone_calloc):
(malloc_zone_free):
(malloc_zone_realloc):
(malloc_zone_memalign):
(malloc_set_zone_name):
(malloc_zone_pressure_relief):
(malloc_zone_print):
* Source/WTF/wtf/malloc_heap_breakdown/malloc/malloc.h: Added.
* Source/WTF/wtf/text/StringImpl.cpp:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/css/parser/CSSParserToken.cpp:
* Source/WebCore/platform/graphics/Region.cpp:
* Source/WebCore/rendering/style/StyleSurroundData.cpp:
* Source/WebKit/CMakeLists.txt:
* Source/bmalloc/CMakeLists.txt:
* Source/bmalloc/bmalloc/BPlatform.h:
* Source/bmalloc/bmalloc/DebugHeap.h:
* Source/bmalloc/bmalloc/IsoTLSInlines.h:
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:
* Source/cmake/WebKitFS.cmake:
* Source/cmake/WebKitFeatures.cmake:
* Tools/Scripts/webkitpy/style/checker.py:
* Tools/TestWebKitAPI/CMakeLists.txt:

Canonical link: <a href="https://commits.webkit.org/294667@main">https://commits.webkit.org/294667@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ae9b7b78cea635e651996e9b9a637461a6a04b3c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102624 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22293 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12607 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107785 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53261 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104663 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22601 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30799 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/78066 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35043 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105630 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17514 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/92621 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58401 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/102099 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/17354 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10652 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52618 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/95295 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/87164 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10720 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110160 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/101233 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29756 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21939 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/87048 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30120 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88817 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/86655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22054 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31475 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9202 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24011 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29684 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34996 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/124867 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29495 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34660 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32818 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31054 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WTF/wtf/PlatformGTK.cmake:
```diff
@@ -1,5 +1,9 @@
 
+if (ENABLE_MALLOC_HEAP_BREAKDOWN)
+    add_subdirectory(glib)
+endif ()
```
We don't use nested CMakeLists.txt below the "WebKit framework" level.

If this code used GLib, then I would say: "paste everything into PlatformWPE.cmake and PlatformGTK.cmake, accepting there will be some code duplication." But it doesn't use GLib or any other Linux platform libraries. Nothing stopping other ports like JSCOnly from using it, right?

So what I would do is put all these build rules directly in Source/WTF/wtf/CMakeLists.txt and just guard it with `if (ENABLE_MALLOC_HEAP_BREAKDOWN AND NOT APPLE)`. Alternatively, you could put it in a named file MallocHeapBreakdown.cmake and include it using `include()` rather than `add_subdirectory()`.

justinmichaud replied:
This makes sense to me

mcatanzaro commented on Source/WTF/wtf/glib/malloc_heap_breakdown/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+    FLATTENED
+)
+
+WEBKIT_FRAMEWORK(MallocHeapBreakdown)
```
`WEBKIT_FRAMEWORK` is for the big projects (WebCore, WTF, JavaScriptCore, WebKit, PAL). It's overkill here. I would create a normal target using `add_library()`.

mcatanzaro commented on Source/WTF/wtf/glib/malloc_heap_breakdown/CMakeLists.txt:
```diff
@@ -0,0 +1,38 @@
+    DESTINATION ${MallocHeapBreakdown_FRAMEWORK_HEADERS_DIR}/malloc
+    FILES ${MallocHeapBreakdown_PUBLIC_HEADERS}
+    FLATTENED
+)
```
Do you really need forwarding headers just for malloc.h? This seems like overkill.

mcatanzaro commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -56,6 +56,7 @@ WEBKIT_OPTION_DEFINE(USE_SKIA_OPENTYPE_SVG "Whether to use the Skia built-in sup
 WEBKIT_OPTION_DEFINE(USE_SOUP2 "Whether to enable usage of Soup 2 instead of Soup 3." PUBLIC OFF)
 WEBKIT_OPTION_DEFINE(USE_SYSTEM_SYSPROF_CAPTURE "Whether to use a system-provided libsysprof-capture" PUBLIC ON)
 WEBKIT_OPTION_DEFINE(ENABLE_JSC_RESTRICTED_OPTIONS_BY_DEFAULT "Whether to enable dangerous development options in JSC by default." PRIVATE OFF)
+WEBKIT_OPTION_DEFINE(ENABLE_MALLOC_HEAP_BREAKDOWN "Whether to enable malloc heap breakdown." PRIVATE OFF)
```
This should go into WebKitFeatures.cmake, since it's not really specific to WPE or GTK ports.

mcatanzaro commented on Source/bmalloc/bmalloc/BPlatform.h:
```diff
@@ -332,7 +332,9 @@
 /* Enable this to put each IsoHeap and other allocation categories into their own malloc heaps, so that tools like vmmap can show how big each heap is. */
+#if !defined(BENABLE_MALLOC_HEAP_BREAKDOWN)
 #define BENABLE_MALLOC_HEAP_BREAKDOWN 0
+#endif
```
```suggestion
#if !defined(BENABLE_MALLOC_HEAP_BREAKDOWN)
#if BOS(DARWIN)
#define BENABLE_MALLOC_HEAP_BREAKDOWN 1
#else
#define BENABLE_MALLOC_HEAP_BREAKDOWN 0
#endif
#endif
```

Scony replied:
that would regress apple ports in the sense that much more code would be compiled

mcatanzaro replied:
OK, I see it's supposed to be off by default.

mcatanzaro commented on Source/bmalloc/bmalloc/DebugHeap.cpp:
```diff
@@ -42,7 +42,7 @@ DebugHeap* debugHeapCache { nullptr };
 DEFINE_STATIC_PER_PROCESS_STORAGE(DebugHeap);
 
-#if BOS(DARWIN)
+#if BENABLE(MALLOC_HEAP_BREAKDOWN) || BOS(DARWIN)
```
This is too confusing. What I would do is guard everything with `BENABLE(MALLOC_HEAP_BREAKDOWN)` only, and always enable it in BPlatform.h if not already defined and `BOS(DARWIN)`. I've left another review comment to suggest this. Then you can have just the single `BENABLE(MALLOC_HEAP_BREAKDOWN)` guard.

justinmichaud replied:
I think Apple folks want this on even when malloc heap breakdown is compiled out. We should make sure we don't regress them, but since we don't support malloc stack logging (as far as I know) on linux, this condition seems correct to me

Scony replied:
I reverted this change to the original one

mcatanzaro commented on Source/cmake/OptionsGTK.cmake:
```diff
@@ -446,6 +447,11 @@ if (APPLE)
 endif ()
 
+if (ENABLE_MALLOC_HEAP_BREAKDOWN)
+  add_definitions(-DBENABLE_MALLOC_HEAP_BREAKDOWN=1)
```
I would do this in Source/bmalloc/CMakeLists.txt.

mcatanzaro reviewed (changes requested):
I'm just reviewing the build system.

I don't see any reason it should be GTK/WPE-specific or go into Source/WTF/wtf/glib. So I would move this up one level and put it directly under WTF/wtf.

mcatanzaro commented on Source/bmalloc/bmalloc/DebugHeap.h:
```diff
@@ -32,7 +32,7 @@
 #include <unordered_map>
 
-#if BOS(DARWIN)
+#if BENABLE(MALLOC_HEAP_BREAKDOWN) || BOS(DARWIN)
```
Should remove the `BOS(DARWIN)` guards here, right? The malloc.h is only wanted when heap breakdown is enabled?

```suggestion
#if BENABLE(MALLOC_HEAP_BREAKDOWN)
```

Scony replied:
I've checked your suggestion on intel-based mac with apple port and it breaks the compilation with default settings (no malloc heap breakdown) - apparently bmalloc's debug heap should be compiled always when `BOS(DARWIN)`

mcatanzaro replied:
It'd be nicer to always build everything on platforms that support it, and guard it behind a runtime setting rather than introduce a new build setting. There's nothing platform-specific in the new Malloc.h, right? It could be used everywhere and gated behind, say, a debug environment variable?

The reason I mention this: we have a long effort to reduce unnecessary build flags, and this pull request is adding a new one.

justinmichaud replied:
In general, I think we should be trying to do the exact opposite; binary size is super important, and you should be able to compile out things you don't need. For linux, I think it is fine to always include this in developer mode builds, so we don't necessarily need a new flag. On macOS though, some but not all of this code is compiled in by default and we shouldn't change the defaults on that platform.

Scony replied:
Well, this particular feature is debug-only and meant to be used very rarely - thus my idea was to have it build-time guarded just like Apple does (see https://docs.webkit.org/Infrastructure/MemoryInspection.html). The only difference here is that the build option is exposed to meta-buildsystem.

As for the new malloc.h - the only platform specific thing about it is that Apple platforms have their own malloc.h. Other than this you're right - new malloc.h does not have anything platform specific inside.

Regarding the new build flag - you mean the cmake part right? So you'd suggest to remove that from cmakes completely and have all the new code (from non-apple platforms) gated behind MALLOC_HEAP_BREAKDOWN macro like apple does?

mcatanzaro replied:
> In general, I think we should be trying to do the exact opposite; binary size is super important,

The WebKit project consensus is that it's not important. WebKit is so massive that cutting a few kilobytes out is just not going to matter, even on tiny embedded devices. But we still receive many bug reports about problems caused by one build flag or another being enabled or disabled, so the cost of the build flags is real and should be avoided whenever possible.

> Regarding the new build flag - you mean the cmake part right? So you'd suggest to remove that from cmakes completely and have all the new code (from non-apple platforms) gated behind MALLOC_HEAP_BREAKDOWN macro like apple does?

I suggest putting it behind a runtime feature flag rather than a build flag. But bmalloc doesn't really have a mechanism for this other than environment variables, and I dislike environment variables. So I guess a build flag is OK. Your call.

mcatanzaro replied:
(If it's going to be a build flag, it's better to expose it to CMake, so that developers don't have to manually edit source code files.)

mcatanzaro replied:
(The usual rule is: if your feature depends on some new external dependency, add a build flag so that use of the external dependency can be disabled. Otherwise: use runtime feature checks only.)

Scony replied:
Well, IMO the build flag also resonates better with current implementation by Apple as `MALLOC_HEAP_BREAKDOWN` macro is already used by Apple ports so by exposing it to cmake as build option, we make Apple's devs lives easier as a side effect.

In that case IMO the current state of PR is what we need.

mcatanzaro replied:
There is value in matching what already exists, sure.

mcatanzaro commented on Source/bmalloc/bmalloc/DebugHeap.h:
```diff
@@ -71,7 +71,7 @@ class DebugHeap : private StaticPerProcess<DebugHeap> {
     static DebugHeap* tryGetSlow();
     
-#if BOS(DARWIN)
+#if BENABLE(MALLOC_HEAP_BREAKDOWN) || BOS(DARWIN)
```
Here too.

Scony replied:
same here

mcatanzaro commented on Source/bmalloc/bmalloc/IsoTLSInlines.h:
```diff
@@ -35,7 +35,7 @@
 #if !BUSE(LIBPAS)
 
-#if BOS(DARWIN)
+#if BENABLE(MALLOC_HEAP_BREAKDOWN)
```
Surely this should be consistent with DebugHeap.h. (I think this is the correct way.)

Scony replied:
fixed

aperezdc reviewed (approved):
I was asked to check the CMake bits, but the rest of the patch LGTM as well 👌🏼 

Of course, the `StringImpl.cpp` conflicts need to be resolved before landing.

Scony committed:
[GTK][WPE] Integrate the ports with MALLOC_HEAP_BREAKDOWN
https://bugs.webkit.org/show_bug.cgi?id=291063

Reviewed by Justin Michaud and Adrian Perez de Castro.

This change integrates GTK and WPE ports with MALLOC_HEAP_BREAKDOWN
feature.

Despite the MALLOC_HEAP_BREAKDOWN was meant to be apple-port only,
it's fairly straightforward for Glib-based ports to export the
malloc.h header along with simple library that implements a mechanism
for tracking the allocations. With that, it's possible to gather
statistics on the current allocations at any point in time.

This change lays a groundwork for more sophisticated integration by
providing a simplified malloc.h header along with a simple library
that allows logging statistics to the standard output periodically.

Co-authored-by: Justin Michaud <jmichaud@igalia.com>

* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/DebugHeap.cpp:
* Source/WTF/wtf/DebugHeap.h:
* Source/WTF/wtf/MallocHeapBreakdown.cmake: Added.
* Source/WTF/wtf/malloc_heap_breakdown/main.cpp: Added.
(MallocZoneHeapManager::getInstance):
(MallocZoneHeapManager::defaultZone):
(MallocZoneHeapManager::createZone):
(MallocZoneHeapManager::renameZone):
(MallocZoneHeapManager::zoneMalloc):
(MallocZoneHeapManager::zoneCalloc):
(MallocZoneHeapManager::zoneRealloc):
(MallocZoneHeapManager::zoneMemalign):
(MallocZoneHeapManager::zoneFree):
(MallocZoneHeapManager::MallocZoneHeapManager):
(MallocZoneHeapManager::~MallocZoneHeapManager):
(MallocZoneHeapManager::monitoringThreadMain):
(malloc_default_zone):
(malloc_create_zone):
(malloc_zone_malloc):
(malloc_zone_calloc):
(malloc_zone_free):
(malloc_zone_realloc):
(malloc_zone_memalign):
(malloc_set_zone_name):
(malloc_zone_pressure_relief):
(malloc_zone_print):
* Source/WTF/wtf/malloc_heap_breakdown/malloc/malloc.h: Added.
* Source/WTF/wtf/text/StringImpl.cpp:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/css/parser/CSSParserToken.cpp:
* Source/WebCore/platform/graphics/Region.cpp:
* Source/WebCore/rendering/style/StyleSurroundData.cpp:
* Source/WebKit/CMakeLists.txt:
* Source/bmalloc/CMakeLists.txt:
* Source/bmalloc/bmalloc/BPlatform.h:
* Source/bmalloc/bmalloc/DebugHeap.h:
* Source/bmalloc/bmalloc/IsoTLSInlines.h:
* Source/cmake/OptionsGTK.cmake:
* Source/cmake/OptionsWPE.cmake:
* Source/cmake/WebKitFS.cmake:
* Source/cmake/WebKitFeatures.cmake:
* Tools/Scripts/webkitpy/style/checker.py:
* Tools/TestWebKitAPI/CMakeLists.txt:

Canonical link: https://commits.webkit.org/294667@main
